content/en/blog/2022-12-05-go-serialization-formats.md

```diff
@@ -0,0 +1,154 @@
+
+```json
+{"name":"Grinch", "age":42, "loc":"Mount Crumpit", "pet":true}
+```
+
+## BSON
+
+BSON Stands for “Binary JSON” and is the format used by MongoDB. BSON is designed to be space efficient, but because BSON encodes extra information like object length in documents, it might be less efficient than JSON in some cases. BSON does this in order to enhance its traversability. BSON is also designed to be transparently converted to/from JSON.
+
+```
+{"bah": "humbug"} →
+\x16\x00\x00\x00            // total document size
+\x02                        // 0x02 = type String
+bah\x00                     // field name
+\x06\x00\x00\x00humbug\x00  // field value
+\x00                        // 0x00 = type EOO ('end of object')
+```
+
+BSON has some special types like "ObjectId", "Min key", and “Max key" which are not compatible with JSON. This incompatibility could lead to some type information getting lost when [converting objects from BSON to JSON](https://stackoverflow.com/questions/12438280/what-is-bson-and-exactly-how-is-it-different-from-json).
+
+
+## MessagePack
+
+MessagePack is used with SignalR, an open source ASP.NET competitor to gRPC, and like protobufs, is specified using a schema. Ironically, MessagePack is more compatible with JSON than BSON, so:
+
+```json
+{"name":"Grinch", "age":42, "loc":"Mount Crumpit", "pet":true}
+```
+becomes...
+```
+DF 00 00 00 04 A4 6E 61 6D 65 A6 47 72 69 6E 63 68 A3 61 67
+65 2A A3 6C 6F 63 AD 4D 6F 75 6E 74 20 43 72 75 6D 70 69 74
+A3 70 65 74 C3
+```
+
+It also supports static-typing and type-checking, which is good news if you are working in a strongly-typed language.
+
+## Protocol Buffers
+
+Protocol Buffers or protobufs are serialization formats used with the Google designed gRPC Remote Procedure Call framework. They are specified with a generic schema that can generate code to be used with a variety of different languages. As mentioned in one of our previous [blog posts](https://rotational.io/blog/what-are-protocol-buffers/), protobufs produce extremely compact messages by using a binary format.
+
+```proto
+syntax = "proto3";
+
+package santa.v1;
+
+import "google/protobuf/timestamp.proto";
+
+message Sleigh {
+    uint64 id = 1;
+    bytes naughty_list = 2;
+    bool lively_and_quick = 3;
+    repeated string reindeer = 4;
+    google.protobuf.Timestamp departure = 5;
+}
+```
+
+## Serialization Bakeoff
+
+Because everyone loves benchmarks, we decided to put each of the serialization formats mentioned above to the test using Golang.
+
+We created three different data structs: small (5 items), medium (20 items), and large (45 items). We were sure to use a range of data types (strings, integers, booleans, and time), but since time can be a little tricky with protocol buffers, we used strings to represent time in this experiment. Finally, we compared JSON, BSON, and MessagePack both in their raw and gzipped form.
```

Comment:
My only request is that you add in a hyperlink for the text "since time can be a little tricky with protocol buffers" to point readers to a more detailed discussion.

Reply:
I amended the sentence to state we chose to use strings over Timestamp and then added a link to the reference about Timestamp.